system/lib/wasmfs/js_file_backend.cpp

```diff
@@ -4,74 +4,24 @@
 // found in the LICENSE file.
 
 // This file defines the JS file backend and JS file of the new file system.
-// Current Status: Work in Progress.
 // See https://github.com/emscripten-core/emscripten/issues/15041.
 
 #include "backend.h"
+#include "js_impl_backend.h"
 #include "wasmfs.h"
 
-using js_index_t = uint32_t;
+// See library_wasmfs_js_file.js
 
 extern "C" {
-int _wasmfs_write_js_file(js_index_t index,
-                          const uint8_t* buffer,
-                          size_t length,
-                          off_t offset);
-int _wasmfs_read_js_file(js_index_t index,
-                         const uint8_t* buffer,
-                         size_t length,
-                         off_t offset);
-int _wasmfs_get_js_file_size(js_index_t index);
-int _wasmfs_create_js_file();
-void _wasmfs_remove_js_file(js_index_t index);
+void _wasmfs_backend_add_js_file(wasmfs::backend_t);
 }
 
 namespace wasmfs {
 
-// This class describes a file that lives in JS Memory
-class JSFile : public DataFile {
-  // This index indicates the location of the JS File in the backing JS array.
-  js_index_t index;
-
-  // JSFiles will write from a Wasm Memory buffer into the backing JS array.
-  __wasi_errno_t write(const uint8_t* buf, size_t len, off_t offset) override {
-    return _wasmfs_write_js_file(index, buf, len, offset);
-  }
-
-  // JSFiles will read from the backing JS array into a Wasm Memory buffer.
-  __wasi_errno_t read(uint8_t* buf, size_t len, off_t offset) override {
-    // The caller should have already checked that the offset + len does
-    // not exceed the file's size.
-    assert(offset + len <= getSize());
-    return _wasmfs_read_js_file(index, buf, len, offset);
-  }
-
-  void flush() override {}
-
-  // The size of the JSFile is defined as the length of the backing JS array.
-  size_t getSize() override { return _wasmfs_get_js_file_size(index); }
-
-public:
-  JSFile(mode_t mode, backend_t backend) : DataFile(mode, backend) {
-    // Create a new file in the backing JS array and store its index.
-    index = _wasmfs_create_js_file();
-  }
-
-  // Remove the typed array file contents in the backing JS array.
-  ~JSFile() { _wasmfs_remove_js_file(index); }
-};
-
-class JSFileBackend : public Backend {
-
-public:
-  std::shared_ptr<DataFile> createFile(mode_t mode) override {
-    return std::make_shared<JSFile>(mode, this);
-  }
-};
-
-// This function is exposed to users to instantiate a new JSBackend.
 extern "C" backend_t wasmfs_create_js_file_backend() {
-  return wasmFS.addBackend(std::make_unique<JSFileBackend>());
+  backend_t backend = wasmFS.addBackend(std::make_unique<JSImplBackend>());
+  _wasmfs_backend_add_js_file(backend);
```

Comment:
One way to get this to be executed on the correct thread would be to call it lazily the first time a backend operation is called, since those operations will be proxied to the worker thread.

An even better way might be to have the proxying backend constructor take a function that creates a backend as its argument rather than having it take a handle to a backend that has already been created. Then the proxying backend constructor can proxy the creation of the underlying backend to the worker thread and the underlying backend can perform initialization in its constructor, as I would have expected.

No matter what we do there, this call should be somewhere in the JSImplBackend so that the C++ API (even if it's private) is able to fully initialize the backend without needing to be called through the C wrapper.

Reply:
Interesting, I like the idea for making the proxied backend get a constructor function. But how would we do that in the C API? Literally a `backend_t (*)()`..? That seems unoptimal but I'm not sure how else.